keps/sig-architecture/4330-compatibility-versions/README.md

```diff
@@ -0,0 +1,1182 @@
+			{Version: mustParseVersion("1.30"), Default: true, PreRelease: Beta},
+		},
+}
+
+type featureOptions struct {
+  AEnabled        bool
+  AHasCapabilityZ bool
+  BEnabled        bool
+  BHandler        func()
+}
+
+func newFeatureOptions(featureGate FeatureGate) featureOptions {
+  opts := featureOptions{}
+  if featureGate.Enabled(FeatureA) {
+    opts.AEnabled = true
+  }
+  if featureGate.CompatibilityVersion() > "1.29" {
+    opts.AHasCapabilityZ = true
+  }
+
+  if featureGate.Enabled(FeatureB) {
+    opts.BEnabled = true
+  }
+  if featureGate.CompatibilityVersion() > "1.28" {
+    opts.BHandler = newHandler
+  } else {
+    opts.BHandler = oldHandler
+  }
+  return opts
+}
+
+// in client.go
+func ClientFunction() {
+  // ...
+  featureOpts := newFeatureOptions(utilfeature.DefaultFeatureGate)
+  if featureOpts.AEnabled {
+    // ...
+    if featureOpts.AHasCapabilityZ {
+      // run CapabilityZ
+    }
+  }
+  if featureOpts.BEnabled {
+    featureOpts.BHandler()
+  }
+  // ...
+}
+
+```
+
+### CEL Environment Compatibility Versioning
+
+CEL environments already [support a compatibility
+version](https://github.com/kubernetes/kubernetes/blob/7fe31be11fbe9b44af262d5f5cffb1e73648aa96/staging/src/k8s.io/apiserver/pkg/cel/environment/base.go#L45).
+The CEL compatibility version is used to ensure when a kubeneretes contol plane
+component reads a CEL expression from storage written by a (N+1) newer version
+(either due to version skew or a rollback), that a compatible CEL environment
+can still be constructed and the expression can still be evaluated.  This is
+achieved by making any CEL environment changes (language settings, libraries,
+variables) available for [stored expressions one version before they are allowed
+to be written by new
+expressions](https://github.com/kubernetes/kubernetes/blob/7fe31be11fbe9b44af262d5f5cffb1e73648aa96/staging/src/k8s.io/apiserver/pkg/cel/environment/environment.go#L38).
+
+The only change we must make for this enhancement is to remove the
+[compatibility version
+constant](https://github.com/kubernetes/kubernetes/blob/7fe31be11fbe9b44af262d5f5cffb1e73648aa96/staging/src/k8s.io/apiserver/pkg/cel/environment/base.go#L45)
+and instead always pass in N-1 of the compatibility version introduced by this
+enhancement as the CEL compatibility version.
+
+### StorageVersion Compatibility Versioning
+
+StorageVersions specify what version an apiserver uses to write resources to etcd
+for each API group. The StorageVersion changes across releases as API groups
+graduate through stability levels.
+
+During upgrades and downgrades, the storage version is particularly important.
+To enable upgrades and rollbacks, the version selected for storage in etcd in 
+version N must be (en/de)codable for k8s versions N-1 through N+1.
+
+Thus, to determine the storage version to use at compatibility version N, we 
+will find the set of all supported GVRs for each of N-1, N, and N+1 and intersect 
```

Comment:
Having to consider N+1 is a little weird here, and I'm not *sure* it's needed.
* We don't know the future; the 1.n binary doesn't have info about 1.(n+1)
* If all versions follow the 1.(n-1) and 1.n rule, then 1.(n+1) would choose a version compatible with 1.n, right?

Reply:
There are planned deprecation already in the code base. 
I imagine there could also be planned feature promotion, when the feature is mostly stable but would like to wait a few cycles to move to the next stage. There could clients would like to try out new features ahead of time.
`binaryVersion=N, compatibilityVersion=N+1` would not behave the same as `binaryVersion=N+1, compatibilityVersion=N+1`, but I think this would be useful specially for testing purpose, to test future compatibility for example.